subsys/net/l2/canbus/canbus.c

```diff
@@ -1,49 +1,1211 @@
+				ret = &tx_ctx[i];
+				break;
+			}
+		}
+	}
+
+	k_mutex_unlock(&tx_ctx_mtx);
+	return ret;
+}
+
+static inline u32_t receive_get_ff_length(struct net_pkt *pkt)
+{
+	u32_t len;
+	u16_t len_u12;
+	int ret;
+
+	ret = net_pkt_read_be16(pkt, &len_u12);
+	if (ret < 0) {
+		NET_ERR("Can't read length");
+	}
+
+	len = len_u12 & 0x0FFF;
+
+	/* Jumbo packet (32 bit length)*/
+	if (!len) {
+		ret = net_pkt_read_be32(pkt, &len);
+	}
+	if (ret < 0) {
+		NET_ERR("Can't 32 bit length");
+	}
+
+	return len;
+}
+
+static inline size_t get_sf_length(struct net_pkt *pkt)
+{
+	u8_t pci;
+	size_t len;
+
+	pci = net_buf_pull_u8(pkt->frags);
+	len = pci & NET_CAN_PCI_SF_DL_MASK;
+
+	/*Single frame with CAN_DL > 8*/
+	if (len == 0) {
+		len = net_buf_pull_u8(pkt->frags);
+	}
+
+	return len;
+}
+
+static inline void set_frame_datalength(struct zcan_frame *frame, u8_t length)
+{
+	/* TODO: Needs update when CAN FD support is added*/
+	NET_ASSERT(length <= NET_CAN_DL);
+	frame->dlc = length;
+}
+
+static enum net_verdict finish_pkt(struct net_pkt *pkt)
+{
+	net_pkt_cursor_init(pkt);
+
+	if (!net_6lo_uncompress(pkt)) {
+		NET_ERR("6lo uncompression failed");
+		return NET_DROP;
+	}
 
 	return NET_CONTINUE;
 }
 
-static inline int canbus_send(struct net_if *iface, struct net_pkt *pkt)
+static inline u32_t addr_to_id(u16_t dest, u16_t src)
+{
+	return (dest << CAN_NET_IF_ADDR_DEST_POS) |
+	       (src  << CAN_NET_IF_ADDR_SRC_POS);
+}
+
+static void set_frame_addr(struct zcan_frame *frame,
+			   const struct net_can_lladdr *dest,
+			   const struct net_can_lladdr *src)
+{
+	frame->id_type = CAN_EXTENDED_IDENTIFIER;
+	frame->rtr = CAN_DATAFRAME;
+
+	frame->ext_id = addr_to_id(dest->addr, src->addr);
+}
+
+
+static void set_frame_addr_pkt(struct zcan_frame *frame, struct net_pkt *pkt,
+			       bool mcast)
+{
+	struct net_can_lladdr src_addr, dest_addr;
+
+	dest_addr.addr = mcast ? NET_CAN_MULTICAST_ADDR : get_dest_lladdr(pkt);
+	src_addr.addr = get_lladdr(net_if_get_link_addr(pkt->iface));
+
+	set_frame_addr(frame, &dest_addr, &src_addr);
+}
+
+static void fc_send_cb(u32_t err_flags, void *arg)
+{
+	if (err_flags) {
+		NET_ERR("Sending FC frame failed: %d", err_flags);
+	}
+}
+
+static int send_fc(struct device *net_can_dev, struct net_can_lladdr *dest,
+		   struct net_can_lladdr *src, u8_t fs)
+{
+	const struct net_can_api *api = net_can_dev->driver_api;
+	struct zcan_frame frame = {
+		.id_type  = CAN_EXTENDED_IDENTIFIER,
+		.rtr = CAN_DATAFRAME,
+	};
+	int ret;
+
+	NET_ASSERT(!(fs & NET_CAN_PCI_TYPE_MASK));
+
+	set_frame_addr(&frame, dest, src);
+
+	frame.data[0] = NET_CAN_PCI_TYPE_FC | fs;
+	/* Block size of 0. This means that all data is sent without waiting for
+	 * additional FC frames
+	 */
+	frame.data[1] = 0;
+	/*STmin*/
+	frame.data[2] = NET_CAN_STMIN;
+	set_frame_datalength(&frame, 3);
+
+	NET_DBG("Sending FC to ID: 0x%08x", frame.ext_id);
+	ret = api->send(net_can_dev, &frame, fc_send_cb, NULL, K_FOREVER);
```

Comment:
I think you could just do `return api->send(net_can_dev, &frame, fc_send_cb, NULL, K_FOREVER);` and avoid having the ret variable.